File: XPile 3-17/Programs/test_concurrency_full.py
# @dr concurrency = threaded

from dataclasses import dataclass, field

def __ul_sort(arr, n): arr[:n] = sorted(arr[:n])

class __ul_Exception(Exception):
    def __init__(self, type_name, message):
        self.type_name = type_name
        self.message = message
        super().__init__(message)

shared_counter = 0

def worker():
    global shared_counter
    i = 0
    while (i < 1000):
        lock(counter_lock)
        shared_counter = (shared_counter + 1)
        unlock(counter_lock)
        i = (i + 1)

def main():
    global shared_counter
    spawn(worker())
    spawn(worker())
    wait()
    print(shared_counter)
    return 0


if __name__ == "__main__":
    main()

File: XPile 3-17/Programs/test_concurrency_full.js
// @dr concurrency = threaded


let shared_counter = 0;
function worker() {
    let i = 0;
    while ((i < 1000)) {
        lock(counter_lock);
        shared_counter = (shared_counter + 1);
        unlock(counter_lock);
        i = (i + 1);
    }
}

function main() {
    spawn(worker());
    spawn(worker());
    wait();
    console.log(shared_counter);
    return 0;
}

function __ul_sort(arr, n) { const sub = arr.slice(0, n).sort((a, b) => a < b ? -1 : a > b ? 1 : 0); for (let i = 0; i < n; i++) arr[i] = sub[i]; }

main();
